### PR TITLE
Expose program, data and mmio read/write operations, for DSP_PDATA/PADR-based DMA-style transfers

### PR DESCRIPTION
**Pull request by PoroCYon**
More stuff melonDS needs, see [here](https://gitlab.ulyssis.org/pcy/melonDS/-/blob/d19962c4e7ea806a07f5e298fdda0e8cdb743793/src/DSi_DSP.cpp#L204-248) for example usage.

### CODE REVIEW

**Comment by wwylele** [comment]
I am not sure if all of this is necessary. Except for MMIO, program/data memory can simply be read/write via `GetDspMemory`.

About `DMAChan0GetDstHigh()` : the address high for D->A data transfer is not the address in the usual DMA engine, but a special register at `ARM_PADDR_HIGH` +0x0186 (See https://github.com/wwylele/teakra/blob/master/src/dma.md). I haven't implemented exposing that register yet.

**Comment by PoroCYon** [comment]
I opted for adding the Program and Data ones as well, as `MemoryInterfaceUnit::DataMemoryOffset` etc. aren't publicly accessible, and hardcoding its value feels dirty.

Re: `ARM_PADDR_HIGH`: hm, interesting, GBATek talks about "pause flags" in that location. But I'll let you double-check that first.

**Comment by wwylele** [comment]
`ARM_PADDR_HIGH`: I checked some old exchange between nocash and me, and it looks like we eventually decided channel0 address is used as D->A address high (= your code is right). I forgot to update the doc on my side. 

> Do you happen to know if setting the DSP_PCFG bits to 7 would make it route back to the ARM/AHB address space in this case as well?

I don't know, but that would be my initial guess.

File: src/memory_interface.h [hunk 1]
```diff
@@ -56,8 +57,12 @@ class MemoryInterface {
     void SetMMIO(MMIORegion& mmio);
     u16 ProgramRead(u32 address) const;
     void ProgramWrite(u32 address, u16 value);
-    u16 DataRead(u16 address); // not const because it can be a FIFO register
-    void DataWrite(u16 address, u16 value);
+    u16 DataRead(u16 address, bool bypass_mmio = false); // not const because it can be a FIFO register
+    void DataWrite(u16 address, u16 value, bool bypass_mmio = false);
+    u16 DataReadA32(u32 address); // not const because it can be a FIFO register
```
Comment (wwylele) [inline src/memory_interface.h:62]: Actually this can't be a FIFO register, right?

Comment (PoroCYon) [reply src/memory_interface.h:62]: You're right, I blindly copied this from the regular `DataRead`. Will fix. (testing the reroute-DMA-back-to-AHB stuff atm)

**Pull request merged**